### DISCUSSION TITLE
Enforce path-style URLs for ETCD backup to S3 bucket

### DISCUSSION THREAD

**Discussion by ostapkostyk** [k3s-io/k3s#9046] [Q&A] [answered]
Hello,
I try to set up the backing up of ETCD of HA cluster to S3-compatible storage. 
I get an error
```
FATA[0008] failed to sync ETCDSnapshotFile: the server is currently unable to handle the request (get etcdsnapshotfiles.meta.k8s.io <some resource>)
```
I don't know what's the problem, but suspect that there might be a problem with URL style. The provider of S3 object storage is Contabo, they do not support virtual-hosted–style URLs.

I follow the [documention](https://docs.k3s.io/cli/etcd-snapshot#s3-compatible-api-support) but there is not stated what kind of URLs does etcd-snapshot use. Also, I don't see an option to choose one.

My command to backup:
```
k3s etcd-snapshot save --s3 --s3-skip-ssl-verify --s3-bucket=etcd-backups --etcd-s3-region=eu2 --etcd-s3-endpoint=$S3_ENDPOINT --etcd-s3-access-key=$S3_ACCESS_KEY --etcd-s3-secret-key=$S3_SECRET_KEY
```

To summarize:
- which URL style does etcd snapshot use to access S3 bucket?
- is there a way to enforce path-style?
- if not, is there some workaround?
- maybe the reason for the error is different, so any other comments are welcome

**Comment by brandond** [comment, marked as answer]
Based on the error message you're getting, that is not the problem. The message suggests there was a problem retrieving the custom resource used to store snapshot metadata, but something odd is going on, as `etcdsnapshotfiles.meta.k8s.io` is not the correct group and version for that resource.

Can you provide the **exact** message that you are getting, along with the output of `kubectl api-resources -o wide` and `kubectl get etcdsnapshotfiles.k3s.cattle.io` ?

**Reply by brandond** [reply]
Also, in the future please open an issue and fill out the issue template.

**Reply by ostapkostyk** [reply]
Thank you very much for your help.
You are right, the problem is not with s3 but with my node. I couldn't get the proper output of ```kubectl api-resources -o wide```, it was complaining about access rights. I was trying to fix that and now K3s service is not running at all on that node. I'd appreciate if you can help me with that as well (shall I continue here or open an issue?).

But I have 2 more healthy nodes in HA cluster, so I tried to backup ETCD to S3 on another one and seems it was successful. I can see a file in my S3 bucket, but the command finished with FATAL error ```FATA[0008] nodes "vmi7.contaboserver.net" not found```. Could you please suggest what can be the problem?

Here is terminal session:
```
me@vmi7:~/s3$ sudo k3s etcd-snapshot save --s3 --s3-bucket=etcd-backups --etcd-s3-region=eu2 --etcd-s3-endpoint=$S3_ENDPOINT --etcd-s3-access-key=$S3_ACCESS_KEY --etcd-s3-secret-key=$S3_SECRET_KEY
INFO[0000] Saving etcd snapshot to /var/lib/rancher/k3s/server/db/snapshots/on-demand-vmi7.contaboserver.net-1702478778
{"level":"info","ts":"2023-12-13T14:46:17.838928Z","caller":"snapshot/v3_snapshot.go:65","msg":"created temporary db file","path":"/var/lib/rancher/k3s/server/db/snapshots/on-demand-vmi7.contaboserver.net-1702478778.part"}
{"level":"info","ts":"2023-12-13T14:46:17.843112Z","logger":"client","caller":"v3@v3.5.9-k3s1/maintenance.go:212","msg":"opened snapshot stream; downloading"}
{"level":"info","ts":"2023-12-13T14:46:17.843191Z","caller":"snapshot/v3_snapshot.go:73","msg":"fetching snapshot","endpoint":"https://127.0.0.1:2379"}
{"level":"info","ts":"2023-12-13T14:46:18.019725Z","logger":"client","caller":"v3@v3.5.9-k3s1/maintenance.go:220","msg":"completed snapshot read; closing"}
{"level":"info","ts":"2023-12-13T14:46:18.023279Z","caller":"snapshot/v3_snapshot.go:88","msg":"fetched snapshot","endpoint":"https://127.0.0.1:2379","size":"4.3 MB","took":"now"}
{"level":"info","ts":"2023-12-13T14:46:18.02344Z","caller":"snapshot/v3_snapshot.go:97","msg":"saved","path":"/var/lib/rancher/k3s/server/db/snapshots/on-demand-vmi7.contaboserver.net-1702478778"}
INFO[0001] Checking if S3 bucket etcd-backups exists
INFO[0001] S3 bucket etcd-backups exists
INFO[0001] Saving etcd snapshot on-demand-vmi7.contaboserver.net-1702478778 to S3
INFO[0001] Uploading snapshot to s3://etcd-backups//var/lib/rancher/k3s/server/db/snapshots/on-demand-vmi7.contaboserver.net-1702478778
INFO[0001] Uploaded snapshot metadata s3://etcd-backups//var/lib/rancher/k3s/server/db/.metadata/on-demand-vmi7.contaboserver.net-1702478778
INFO[0001] S3 upload complete for on-demand-vmi7.contaboserver.net-1702478778
INFO[0002] Reconciling ETCDSnapshotFile resources
INFO[0002] Creating ETCDSnapshotFile for etcd-snapshot-contabo-2-1702382403
INFO[0003] Creating ETCDSnapshotFile for etcd-snapshot-contabo-2-1702425605
INFO[0003] Creating ETCDSnapshotFile for etcd-snapshot-contabo-2-1702468805
INFO[0004] Creating ETCDSnapshotFile for etcd-snapshot-contabo-2-1702296000
INFO[0004] Creating ETCDSnapshotFile for etcd-snapshot-contabo-2-1702339204
INFO[0008] Reconciliation of ETCDSnapshotFile resources complete
FATA[0008] nodes "vmi7.contaboserver.net" not found
me@vmi7:~/s3$ sudo kubectl get nodes -o wide
NAME        STATUS     ROLES                       AGE     VERSION        INTERNAL-IP     EXTERNAL-IP     OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
contabo-1   NotReady   control-plane,etcd,master   5d1h    v1.27.7+k3s2   EXT_IP_1        EXT_IP_1        Ubuntu 22.04.3 LTS   5.15.0-25-generic   containerd://1.7.7-k3s1.27
contabo-2   Ready      control-plane,etcd,master   5d1h    v1.27.7+k3s2   EXT_IP_2    	  EXT_IP_2        Ubuntu 22.04.3 LTS   5.15.0-25-generic   containerd://1.7.7-k3s1.27
contabo-3   Ready      control-plane,etcd,master   4d20h   v1.27.7+k3s2   EXT_IP_3        EXT_IP_3        Ubuntu 22.04.3 LTS   5.15.0-25-generic   containerd://1.7.7-k3s1.27
```
I edited the node name to "vmi7" for privacy, but this is actually the name of the node where I ran that session (the node's own name). I mean, this is not the name of my other unhealthy node. 

When I installed K3s, I changed the name of the node with the ```--node-name``` option (sorry, I masked real names and IPs):
```
curl -sfL https://get.k3s.io | sh -s - server --server https://NAME_1:6443 --token=$K3S_TOKEN --node-name=NAME_2 --node-external-ip=EXT_IP_2 --flannel-backend=wireguard-native --flannel-external-ip
```
maybe this is somehow related to the problem

some logs:
[k3s_log1.txt](https://github.com/k3s-io/k3s/files/13662332/k3s_log1.txt)
[k3s_log3.txt](https://github.com/k3s-io/k3s/files/13662674/k3s_log3.txt)

**Reply by brandond** [reply]
If you change the node-name in the `k3s server` args, you should also pass `--node-name` when running `k3s etcd-snapshot`. Or better yet, just put `node-name: NAME_2` in /etc/rancher/k3s/config.yaml and it will always be set correctly.
